[RNMobile] Resolve Text Entry Issues/Errors in Alt Text Settings

SiobhyB opened the pull request:
**Fixes:** 

* https://github.com/WordPress/gutenberg/issues/33857
* https://github.com/WordPress/gutenberg/issues/33856
* https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368

**Gutenberg Mobile:**

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/3818

## Description

The code introduced in this PR addresses the following three issues with the text entry field in the image block's alt text settings:

* The cursor would jump to different positions at times when typing swiftly or deleting text on Android.
* Text would duplicate itself when typing swiftly in both Android and iOS.
* The following error was logged in the developer console when opening the image block's settings:

```
ExceptionsManager.js:180 Warning: Cannot update a component (`Unknown`) while rendering a different component (`BottomSheetTextControl`). To locate the bad setState() call inside `BottomSheetTextControl`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
    in BottomSheetTextControl (at edit.native.js:416)
```

An open issue in the React Native repository, https://github.com/facebook/react-native/issues/30503, lines up with the behavior described in the first two issues. That issue outlines the fact that this was introduced in React Native 0.63.0 and is related to global state management with the `TextInput` component. The third issue is semi-related, and resolved through the code from this PR, given it also touches on state management.

Further technical details around the code changes are provided in the **Types of changes** section below.

## How has this been tested?

I completed the following test cases on a physical **Pixel 5, Android version 11** and **an iPhone XR, iOS 14.6**. 

While going through the below flows, it'd be useful to also think about and experiment with different edge cases e.g. entering return at random while typing swiftly or clicking delete abruptly.

### Test Case 1: General Flow

<details>
<summary>To start with, the general flow for adding new alt text is as follows.</summary>

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post or open an existing one. Add a new image block and upload an image to it.
* Tap the image block's cog/gear icon to access its settings. 
* Tap **Alt Text** to access the text editor for adding alt text to the image.
* Enter new text into the text editor and exit the settings.
* Confirm that the image's alt text has been correctly updated. This can be further verified by saving changes and confirming the alt text has been correctly updated on the web.

</details>

### Test Case 2: Deleting Text

<details>
<summary>We should also verify that the cursor no longer changes position at random after deleting text.</summary>

* Follow the first three steps in the **General Flow** test case to access the alt text settings within an image block's settings panel.
* After entering text into the editor, move the cursor somewhere in the middle and press the delete/backspace key.
* Verify that one character is deleted and the cursor doesn't change position. 
</details>

### Test Case 3: Typing Swiftly

<details>
<summary>Next, we should verify that it's possible to type swiftly into the text editor, with no unwanted side effects, such as duplication.</summary>

* Again, follow the first three steps in the **General Flow** test case to access the alt text settings within an image block's settings panel.
* Attempt to enter a sentence describing the image very quickly into the editor. 
* Verify that the text you've entered appears correctly in the editor, with no duplication or errors. 
</details>

### [Developer] Test Case 4: Verify No Errors in Developer Console

While testing the branch on a local setup, confirm that there is no longer a warning related to the `BottomSheetTextControl` component when opening the image block's settings.

## Screenshots

Previously, hitting the backspace/delete key would lead to the cursor jumping, but you'll see this is no longer the case in the following video:

https://user-images.githubusercontent.com/2998162/128915167-4389a02c-94a5-4292-871d-b07f2968a607.mov

## Types of changes

This PR introduces a bug fix (a non-breaking change which fixes an issue) with the following high-level overview of the code changes:

* React Native's `TextInput` component is being used in [the `BottomSheetTextControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet-text-control/index.native.js). It's this component that's used to enable users to enter alt text via the image block's settings.
* The `onChange` prop was being used to create a side effect when text was typed into the `TextInput` component. As [`onChange` uses `setState`](https://github.com/facebook/react/issues/18178#issuecomment-596880684), this was creating [the error outlined here](https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368). The solution was to [remove the reference to `onChange`](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eL81) and [move the functionality to a `useEffect` hook](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eR47).
* In addition to the above changes, the `TextInput` component's `value` prop was [changed to `defaultValue`](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eR85), as per [the suggestion in this open bug report](https://github.com/facebook/react-native/issues/30503#issuecomment-787001771) to workaround a known React Native issue with the `TextInput` component. 

<hr />

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Resolve render error

This commit aims to fix the render error outlined here: https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368

The "outside" changes that happen when text is changed in the BottomSheetTextControl component should not be called directly from that component. Instead, they're now called within the useEffect hook.

SiobhyB committed:
Replace "value" prop with "defaultValue"

The purpose of this commit is to address the text entry issues that currently exist for the TextInput component, as described here: https://github.com/facebook/react-native/issues/30503

A working workaround propose in that thread is to replace the "value" prop with "defaultValue".

SiobhyB committed:
Merge branch 'trunk' into rnmobile/fix/bottomsheet-text-control-error

SiobhyB commented:
@fluiddot, looking into a fix for issues with text entry in the alt text settings also had the happy side effect of addressing the error you first reported at https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368. :) I created installable builds for Android and iOS to help with testing, as I think it's easier to both replicate the issues and confirm the fix on physical devices. Let me know if I can help to clarify anything on this PR!

fluiddot commented:
> @fluiddot, looking into a fix for issues with text entry in the alt text settings also had the happy side effect of addressing the error you first reported at [#33057 (review)](https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368). :) I created installable builds for Android and iOS to help with testing, as I think it's easier to both replicate the issues and confirm the fix on physical devices. Let me know if I can help to clarify anything on this PR!

Awesome, thanks @SiobhyB for addressing all the fixes here and create the installable builds 🎊 ! I'll try to spare some time this week to take a look and let you know if I have any feedback 🙇 .

fluiddot commented on packages/components/src/mobile/bottom-sheet-text-control/index.native.js:
```diff
@@ -78,8 +82,7 @@ const BottomSheetTextControl = ( {
 					<TextInput
 						label={ label }
 						onChangeText={ ( text ) => onChangeText( text ) }
-						onChange={ onChange( value ) }
-						value={ value }
+						defaultValue={ value }
```
Looks like that using `defaultValue` will make it unnecessary to keep the value in the state. I think we could simplify this component if we use the incoming prop `initialValue` for the `defaultValue` prop and for the `onChangeText` we could invoke the prop `onChange` directly.

**Changes:**
- `defaultValue={ value }` => `defaultValue={ initialValue }`
- Remove `const [ value, onChangeText ] = useState( initialValue );`
- `onChangeText={ ( text ) => onChangeText( text ) }` => `onChangeText={ ( text ) => onChange( text ) }`

wdyt?

SiobhyB replied:
That works perfectly, thank you for the suggestion and clear explanation of why it works! I was also able to remove the `useEffect` hook with the necessary changes and have implemented that in https://github.com/WordPress/gutenberg/pull/33975/commits/ccbef5cba209116b7b95fe7ac518b81611b706de.

fluiddot reviewed (commented):
Thanks @SiobhyB for addressing the issues 💯 !

I verified the different test cases but I posted a comment regarding the possibility to simplify the component, let me know your opinion.

- Test Case 1: General Flow ✅ 
- Test Case 2: Deleting Text ✅ 
- Test Case 3: Typing Swiftly ✅ 
- [Developer] Test Case 4: Verify No Errors in Developer Console ✅ 

_Tested on Simulator iPad Pro (12.9-inch) - 4th generation (iOS 14.4) and Samsung Galaxy S20 FE 5G (Android 10)._

SiobhyB committed:
Merge branch 'trunk' into rnmobile/fix/bottomsheet-text-control-error

SiobhyB committed:
Simplify component by removing unecessary hooks

The introduction of the "defaultValue" prop in https://github.com/WordPress/gutenberg/pull/33975/commits/f6d9697ab66a05e43d253f853de8ec74f12a73b6 means that it's no longer necessary to keep the value in state. "defaultValue" allows for text to be automatically changed as it's typed. See: https://reactnative.dev/docs/textinput#defaultvalue

We don't need to use "onChangeText" for the purpose of updating the input's text and can instead use it to invoke the "onChange" function.

We can also remove the useState and useEffect hooks, as they're not necessary to update the component's value.

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested on Simulator iPhone 11 Pro Max (iOS 14.4) and Samsung Galaxy S20 FE 5G (Android 10)._

I tested all cases again and they worked perfect 💯 :

- Test Case 1: General Flow ✅
- Test Case 2: Deleting Text ✅
- Test Case 3: Typing Swiftly ✅
- [Developer] Test Case 4: Verify No Errors in Developer Console ✅
